### PR TITLE
Add units='step'

### LINKED ISSUE #2169: Method for "wait minimum time"
Following the discussion in https://github.com/cocotb/cocotb/pull/2153#pullrequestreview-521152671, we should have a method like ``Timer.shortest()`` to more semantically express that a minimum waiting time is wanted.

### LINKED ISSUE #2171: Time unit "step" should represent the simulator time step
Right now, ``Timer`` and ``utils.get_sim_steps`` use ``units=None`` (the default value) to mean "use the simulator time step".
Add ``units='step'`` (as exists in SystemVerilog) to mean the same.

### PR DESCRIPTION
**Pull request by cmarqu**
Use Timer.shortest() in cocotb_test.

Closes #2169, closes #2171

### CODE REVIEW

**Comment by cmarqu** [comment]
TBD: add test in ``test_cocotb/test_timing_triggers.test_timer_with_units``.

**Comment by cmarqu** [comment]
TBD: Check if docs need to be updated.

**Review by ktbarrett — COMMENTED** [review]
I'm not sure if I like having a 'shortest' shortcut, it's not really offering anything over just writing `Timer(1, 'step')`.

**Comment by cmarqu** [comment]
Ok, I could just make a reference to such a Timer with a speaking name in the tests without offering it as part of cocotb.

**Comment by cmarqu** [comment]
Tbd: add versionchanged tag

**Comment by alexforencich** [comment]
Maybe there is an argument for having `Timer.ns(n)`, `Timer.step(1)`, etc.

**Comment by ktbarrett** [comment]
@alexforencichI agree I don't like the string arguments. I had started #1130 for making a physical type for simulation time that worked like some chrono libraries.

**Comment by cmarqu** [comment]
TBDs fixed, doc best found via https://cocotb--2172.org.readthedocs.build/en/2172/release_notes.html#features

File: cocotb/utils.py [hunk 1]
```diff
@@ -118,7 +118,7 @@ def get_sim_steps(time, units=None):
         :exc:`ValueError`: If given *time* cannot be represented by simulator precision.
     """
     result = time
-    if units is not None:
+    if units is not None and units != "step":
```
Comment (ktbarrett) [inline cocotb/utils.py:121]: This is the typical idiom. 
```suggestion
    if units not in (None, 'step'):
```
One could also do a default check to correct `None` to `'step'` and only do the singular check  of `'step'` thereafter.

File: cocotb/clock.py [hunk 2]
```diff
@@ -66,10 +66,9 @@ class Clock(BaseClock):
         period (int): The clock period. Must convert to an even number of
             timesteps.
         units (str, optional): One of
-            ``None``, ``'fs'``, ``'ps'``, ``'ns'``, ``'us'``, ``'ms'``, ``'sec'``.
-            When no *units* is given (``None``) the timestep is determined by
-            the simulator.
-
+            ``None``, ``'step'``, ``'fs'``, ``'ps'``, ``'ns'``, ``'us'``, ``'ms'``, ``'sec'``.
+            When no *units* is given (``None``) or is ``'step'``,
```
[the remark this answers is not in the document]
Comment (cmarqu) [reply cocotb/clock.py:70]: Yep, thanks.

File: cocotb/clock.py [hunk 3]
```diff
@@ -66,10 +66,9 @@ class Clock(BaseClock):
...
+            the timestep is determined by the simulator.
```
[the remark this answers is not in the document]
Comment (cmarqu) [reply cocotb/clock.py:71]: I did the linking. I don't feel it's bad that we don't explain details here, the user just needs to pick a unit that makes sense to them. "Not sure what your simulator's time step is? Don't rely on that then."

**Comment by garmin-mjames** [comment]
Newsfragment should be 2172, not 2171.

**Comment by cmarqu** [comment]
@garmin-mjames Towncrier works fine with either issue or PR, and in this case, the first item in the PR is misleading because obsolete (and editing it now feels wrong).
We have linked to issues in the past, they actually explain the problem being solved better most of the time (but there are PRs without accompanying issues, so we link to the PR then).

File: cocotb/utils.py [hunk 4]
```diff
@@ -118,7 +118,7 @@ def get_sim_steps(time, units=None):
         :exc:`ValueError`: If given *time* cannot be represented by simulator precision.
     """
     result = time
-    if units is not None:
+    if units not in (None, "step"):
```
Comment (eric-wieser) [inline cocotb/utils.py:121]: Is there any point supporting `None` here? It seems odd to introduce two ways to spell the same thing, unless our plan is to deprecate one of them.

I'd be tempted to change the default to `units="step"`.

File: tests/test_cases/test_cocotb/test_timing_triggers.py [hunk 5]
```diff
@@ -39,35 +39,39 @@ async def test_function_reentrant_clock(dut):
...
 
+    time_fs = get_sim_time(units='fs')
+    await Timer(1, units='step')
+    assert time_step == get_sim_time(units='fs') - time_fs, "Expected units='step' == units=None"
+
     try:
         # Await for 2.5 timesteps, should throw exception
         await Timer(2.5*time_step, units='fs')
-        raise TestFailure("Timers should throw exception if time cannot be achieved with simulator resolution")
+        assert False, "Timers should throw exception if time cannot be achieved with simulator resolution"
```
Comment (eric-wieser) [inline tests/test_cases/test_cocotb/test_timing_triggers.py:53]: If you're changing this, may as well use `assert_raises`

File: cocotb/utils.py (hunk 4 above)
Comment (ktbarrett) [reply cocotb/utils.py:121]: This change will need to be reflected in the docstring and newsfrag, as a reminder.

Comment (cmarqu) [reply cocotb/utils.py:124]: Changed the default and deprecated None, with tests. Docstrings and newsfragments should be fine now too.

File: tests/test_cases/test_cocotb/test_timing_triggers.py (hunk 5 above)
Comment (cmarqu) [reply tests/test_cases/test_cocotb/test_timing_triggers.py]: Done.

File: documentation/source/newsfragments/2171.feature.rst [hunk 6]
```diff
@@ -0,0 +1,6 @@
+The ``units`` argument to :class:`cocotb.triggers.Timer`,
+:class:`cocotb.clock.Clock` and :func:`cocotb.utils.get_sim_steps`,
+and the ``timeout_unit`` argument to
+:func:`cocotb.triggers.with_timeout` and :class:`cocotb.test`
+now accepts ``'step'`` to mean the simulator time step.
+This used to be expressed using ``None`` which is now deprecated.
```
Comment (ktbarrett) [inline documentation/source/newsfragments/2171.feature.rst:6]:
```suggestion
This used to be expressed using ``None``, which is now deprecated.
```

**Review by ktbarrett — APPROVED** [review]

**Pull request merged**